### PR TITLE
Fetch all authors from REST API and make dropdown scrollable

### LINKED ISSUE #756: Author dropdown does not scroll.
## Bug Description

If the site has more than 10 users, the dropdown doesn't have a scrollbar. 

## Expected Behaviour

The author dropdown should scroll or better it should be an autocomplete search input. 

## Steps to Reproduce
Tip to test locally. 
Run the following command. 
```docker-compose -f bin/local-env/docker-compose.yml exec cli wp user generate --count=200```

- Go to editor. 
- Select document sidebar. 
- Select author dropdown. 

## Screenshots

Screenshot from 2020-03-25 13-05-42

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by ndev1991**
Issue #756 

Add scrollbar inside dropdown and set the max-height.
Screen Shot 2020-03-26 at 2 10 41 PM

### CODE REVIEW

**Comment by barklund** [comment]
I think the larger issue here is, that only 10 users are loaded from the db. All db users should be loaded and selectable in the dropdown.

CC @spacedmonkey

**Comment by spacedmonkey** [comment]
I think the issue here is that, the REST API only returns 10 by default. The api maxes out at 100. As to get all users, we would have to pagination through all users.

**Comment by ndev1991** [comment]
@barklund @spacedmonkey So what should we accomplish from this issue?

File: assets/src/edit-story/app/api/apiProvider.js [hunk 1]
```diff
@@ -240,9 +240,30 @@ function APIProvider({ children }) {
...
+      apiPath = addQueryArgs(apiPath, {
+        per_page: perPage,
+        page: pagingNum,
+      });
+      return apiFetch({ path: apiPath });
+    },
+    [users]
+  );
+
+  const getAllUsers = useCallback(
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/api/apiProvider.js]: We might be able to get the results quicking by using the `X-WP-TotalPages`, [see](https://github.com/google/web-stories-wp/blob/master/assets/src/edit-story/app/media/mediaProvider.js#L73). Get the count of local pages and perform all the api requests at once. Did you try this when building this?

Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: No, I didn't tried it. I will try it now

Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: Yeah, now refactored and uses Promise.all once for all pages.

File: assets/src/edit-story/app/api/apiProvider.js [hunk 2]
```diff
@@ -240,9 +240,46 @@ function APIProvider({ children }) {
...
+        }
+        return Promise.all(promises).then((response) => {
+          return data.concat(
+            response.reduce(
+              (resultArray, result) => resultArray.concat(result.data),
+              []
+            )
+          );
+        });
+      });
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/api/apiProvider.js]: Don't we need to order the array now. What if page 5 return before page 4? This promises, you don't know what order they will be fulfilled

Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: hmm... Good point. Let me update it! 👍

Comment (spacedmonkey) [reply assets/src/edit-story/app/api/apiProvider.js]: It should be sorted name field in the api.

Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: We need sort param as well? Or you mean it supposed to be sorted by name originally?

Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: Oh, it's original feature!

Comment (spacedmonkey) [reply assets/src/edit-story/app/api/apiProvider.js]: I don't think it needs a sort param for now. I think before alpha we should change this to an autocomplete box. But for now, let's just go with sorting by name.

Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: Done!

File: assets/src/edit-story/app/api/apiProvider.js [hunk 3]
```diff
@@ -240,9 +240,47 @@ function APIProvider({ children }) {
     return apiFetch({ path });
   }, [statuses]);
 
-  const getAllUsers = useCallback(() => {
-    return apiFetch({ path: users });
-  }, [users]);
+  const getUsersByPage = useCallback(
+    (pagingNum = 1) => {
+      let apiPath = users;
+      const perPage = 10;
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/api/apiProvider.js]: I think that still should remain 100. On my local dev, I have 564 users and this ended up in 56 requests. Which feels a like crazy. It also took a lot longer to resolve.

Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: Yes, it should be 100! And I've got 108 users and got successfully.

File: assets/src/edit-story/app/api/apiProvider.js [hunk 4]
```diff
@@ -240,9 +240,47 @@ function APIProvider({ children }) {
...
+
+          return { data: jsonArray, headers: response.headers, pagingNum };
+        }
+      );
+    },
+    [users]
+  );
+
+  const getAllUsers = useCallback(
+    (pagingNum = 1) => {
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/api/apiProvider.js]:
```suggestion
    () => {
```

File: assets/src/edit-story/app/api/apiProvider.js [hunk 5]
```diff
@@ -240,9 +240,47 @@ function APIProvider({ children }) {
...
+      return getUsersByPage(pagingNum).then(({ data, headers }) => {
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/api/apiProvider.js]:
```suggestion
      return getUsersByPage().then(({ data, headers }) => {
```

File: assets/src/edit-story/app/api/apiProvider.js (hunk 4 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/api/apiProvider.js]: This param is not used for anything.

**Comment by spacedmonkey** [comment]
Now we have loading from multiple api request, it take a look time for the api requests to load. Should we check if `users` is not empty on [this](https://github.com/google/web-stories-wp/blob/9c42a748722590880ca72cf4099adcd024262d0c/assets/src/edit-story/components/inspector/document/publish/index.js#L195-L207) line now?

**Comment by spacedmonkey** [comment]
Screenshot from 2020-03-27 17-32-07
Shouldn't this dropdown be wider?
I am looking at this in chrome and firefox on linux.

**Comment by ndev1991** [comment]
> Now we have loading from multiple api request, it take a look time for the api requests to load. Should we check if `users` is not empty on [this](https://github.com/google/web-stories-wp/blob/9c42a748722590880ca72cf4099adcd024262d0c/assets/src/edit-story/components/inspector/document/publish/index.js#L195-L207) line now?

I think `users` is null at the beginning and updated when there is values. And also we have checking `users` like `&& users &&`.

**Comment by spacedmonkey** [comment]
There is already a check, but I am seeing an empty dropdown for a bit.

**Comment by ndev1991** [comment]
> There is already a check, but I am seeing an empty dropdown for a bit.

Okay. Then we may need that to dropdown.

**Comment by spacedmonkey** [comment]
Should be check if see if not not null users and length is more than 0?

**Comment by ndev1991** [comment]
> Should be check if see if not not null users and length is more than 0?

Yes, basically we have checking point for `users` not null for that part. And we will need to add length check to dropdown I think.

**Comment by ndev1991** [comment]
> > Should be check if see if not not null users and length is more than 0?
> 
> Yes, basically we have checking point for `users` not null for that part. And we will need to add length check to dropdown I think.

hmm... that maybe doesn't make sense. It would be good to add length check after ` && users && `

**Comment by ndev1991** [comment]
@spacedmonkey basically if we add that beside users check in then the author line with dropdown will appear after all the users loaded. And if we add that to dropdown then the author line shows and the correct value set for dropdown after users fully loaded. What do you think about this? Which would be the expected behavior for this?

**Comment by spacedmonkey** [comment]
> @spacedmonkey basically if we add that beside users check in then the author line with dropdown will appear after all the users loaded. And if we add that to dropdown then the author line shows and the correct value set for dropdown after users fully loaded. What do you think about this? Which would be the expected behavior for this?

The user dropdown loads after in gutenberg. The alternative here to a dropdown that doesn't work at all. The best UX would be only load the users when the dropdown is clicked, as they may never use the dropdown at all. But that is more work. 

CC @pbakaus @swissspidy What do you think? Hide the author dropdown until all users have loaded or user something and have a possibly none functioning dropdown for a while.

**Comment by pbakaus** [comment]
> > @spacedmonkey basically if we add that beside users check in then the author line with dropdown will appear after all the users loaded. And if we add that to dropdown then the author line shows and the correct value set for dropdown after users fully loaded. What do you think about this? Which would be the expected behavior for this?
> 
> The user dropdown loads after in gutenberg. The alternative here to a dropdown that doesn't work at all. The best UX would be only load the users when the dropdown is clicked, as they may never use the dropdown at all. But that is more work.
> 
> CC @pbakaus @swissspidy What do you think? Hide the author dropdown until all users have loaded or user something and have a possibly none functioning dropdown for a while.

I would definitely recommend loading it async (not on demand when clicking the dropdown, but also not in the critical path of the editor).

So:

1. on load, show a placeholder for the dropdown ("Loading..", reserving the same space so it doesn't jump around of course)
2. After the editor is fully initialized, start loading the authors
3. Replace placeholder with dropdown.

Relatedly, bonus points for loading the entire JS for this async, and not have a single bundle for everything.

**Comment by spacedmonkey** [comment]
@ndev1991 Did you try passing `-1` to the api?

See this comment. https://github.com/WordPress/gutenberg/issues/21212#issuecomment-605290187

**Comment by ndev1991** [comment]
> @ndev1991 Did you try passing `-1` to the api?
> 
> See this comment. [WordPress/gutenberg#21212 (comment)](https://github.com/WordPress/gutenberg/issues/21212#issuecomment-605290187)

@spacedmonkey yes. It's same with 100. Returns 100 users at a time when using -1

File: assets/src/edit-story/app/api/apiProvider.js (hunk 4 above)
Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: Fixed

File: assets/src/edit-story/app/api/apiProvider.js (hunk 5 above)
Comment (ndev1991) [reply assets/src/edit-story/app/api/apiProvider.js]: Fixed

File: assets/src/edit-story/components/inspector/document/publish/index.js [hunk 6]
```diff
@@ -93,7 +93,7 @@ const StyledToggleIcon = styled(ToggleIcon)`
 
 function PublishPanel() {
   const {
-    state: { users },
+    state: { users, isUsersLoading },
```
Comment (spacedmonkey) [inline assets/src/edit-story/components/inspector/document/publish/index.js:96]: Where is `isUsersLoading` setup and set?

Comment (ndev1991) [reply assets/src/edit-story/components/inspector/document/publish/index.js:96]: @spacedmonkey This logic is on `InspectorProvider` where getAllUsers function called.

Comment (spacedmonkey) [reply assets/src/edit-story/components/inspector/document/publish/index.js:96]: Fair enough.

**Review by spacedmonkey — APPROVED** [review]
LGTM

**Comment by swissspidy** [comment]
Just to clarify: passing `-1` didn't work, despite the [`fetchAllMiddleware`](https://github.com/WordPress/gutenberg/blob/f322b4483974278f72257266cf69944b051b08ea/packages/api-fetch/src/middlewares/fetch-all-middleware.js) effectively doing the same thing as this PR? 🤔

**Comment by ndev1991** [comment]
> Just to clarify: passing `-1` didn't work, despite the [`fetchAllMiddleware`](https://github.com/WordPress/gutenberg/blob/f322b4483974278f72257266cf69944b051b08ea/packages/api-fetch/src/middlewares/fetch-all-middleware.js) effectively doing the same thing as this PR? 🤔

@swissspidy Yes, passing `-1` didn't work. It's same as `per_page=100`. But not tested fetchAllMiddleware

**Comment by swissspidy** [comment]
@ndev1991 Please test this again. The following code automatically resulted in multiple HTTP requests being made, loading all users on the site:

```js
const getAllUsers = useCallback(() => {
    return apiFetch({ path: addQueryArgs(users, { per_page: '-1' })});
  }, [users]);
```

**Review by swissspidy — CHANGES_REQUESTED** [review]
In my testing this just worked for me:

```js
const getAllUsers = useCallback(() => {
  return apiFetch({ path: addQueryArgs(users, { per_page: '-1' })});
}, [users]);
```

Because api-fetch automatically handles this for us.

**Comment by ndev1991** [comment]
> In my testing this just worked for me:
> 
> ```js
> const getAllUsers = useCallback(() => {
>   return apiFetch({ path: addQueryArgs(users, { per_page: '-1' })});
> }, [users]);
> ```
> 
> Because api-fetch automatically handles this for us.

@swissspidy yes it's working. I didn't see it makes multiple http request to fetch all the users.
Updated the code and pushed! Thanks

**Review by swissspidy — APPROVED** [review]

**Pull request merged**